### PR TITLE
Destination Snowflake: Improve error handling in StagingClient

### PR DESCRIPTION
**Pull request by gisripa**
## What
Use the returned results to determine success/failure for  `PUT` and `COPY INTO` commands.

## Can this PR be safely reverted and rolled back?

- [x] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Comment by gisripa** [comment]
* **#39107** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/39107?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#39135** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/39135?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @gisripa and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: airbyte-integrations/connectors/destination-snowflake/src/main/kotlin/io/airbyte/integrations/destination/snowflake/operation/SnowflakeStagingClient.kt [hunk 1]
```diff
@@ -84,7 +111,8 @@ class SnowflakeStagingClient(private val database: JdbcDatabase) {
             filePath,
             stageName,
             stagingPath,
-            Runtime.getRuntime().availableProcessors()
+            // max allowed param is 99, we don't need so many threads for a single file upload
+            minOf(Runtime.getRuntime().availableProcessors(), 4)
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-snowflake/src/main/kotlin/io/airbyte/integrations/destination/snowflake/operation/SnowflakeStagingClient.kt:115]: why not `minOf(..., 99)`? wouldn't this restrict us to 4 threads in most cases?

File: airbyte-integrations/connectors/destination-snowflake/src/test-integration/kotlin/io/airbyte/integrations/destination/snowflake/operation/SnowflakeStagingClientIntegrationTest.kt [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    private var namespace: String = ""
+    private var tablename: String = ""
+
+    private lateinit var stageName: String
+    private val config =
+        Jsons.deserialize(Files.readString(Paths.get("secrets/1s1t_internal_staging_config.json")))
+    private val datasource =
+        SnowflakeDatabaseUtils.createDataSource(config, OssCloudEnvVarConsts.AIRBYTE_OSS)
+    private val database: JdbcDatabase = SnowflakeDatabaseUtils.getDatabase(datasource)
+    // Intentionally not using actual columns, since the staging client should be agnostic of these
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-snowflake/src/test-integration/kotlin/io/airbyte/integrations/destination/snowflake/operation/SnowflakeStagingClientIntegrationTest.kt:42]: random thought: do we eventually want to explicitly specify column names in our `COPY` command? 😅

**Review by edgao — APPROVED** [review]
is SnowflakeStagingClientTest still useful, now that there's a proper integration test?

(two minor comments, neither blocking)

File: airbyte-integrations/connectors/destination-snowflake/src/test-integration/kotlin/io/airbyte/integrations/destination/snowflake/operation/SnowflakeStagingClientIntegrationTest.kt (hunk 2 above)
Comment (gisripa) [reply airbyte-integrations/connectors/destination-snowflake/src/test-integration/kotlin/io/airbyte/integrations/destination/snowflake/operation/SnowflakeStagingClientIntegrationTest.kt:42]: I guess so.. it requires `COPY INTO .. select filename#colindex` or something i think right ?

File: airbyte-integrations/connectors/destination-snowflake/src/main/kotlin/io/airbyte/integrations/destination/snowflake/operation/SnowflakeStagingClient.kt (hunk 1 above)
Comment (gisripa) [reply airbyte-integrations/connectors/destination-snowflake/src/main/kotlin/io/airbyte/integrations/destination/snowflake/operation/SnowflakeStagingClient.kt:115]: in our cloud case it is always 1, our pod config is 1 cpu i believe based on my previous dev release run.

**Comment by gisripa** [comment]
> is SnowflakeStagingClientTest still useful, now that there's a proper integration test?
> 
> (two minor comments, neither blocking)

Yeah not useful for happy path, only useful part was if the correct exception caught and thrown as ConfigError.

File: airbyte-integrations/connectors/destination-snowflake/src/main/kotlin/io/airbyte/integrations/destination/snowflake/operation/SnowflakeStagingClient.kt (hunk 1 above)
Comment (talnidam) [reply airbyte-integrations/connectors/destination-snowflake/src/main/kotlin/io/airbyte/integrations/destination/snowflake/operation/SnowflakeStagingClient.kt:115]: @gisripa 
1. The pod config could also run on 2 cpus by values.yaml definition ain't im wrong? (worker resources)
2. The update will be released on the next snowflake dest version? or the current one?

File: airbyte-integrations/connectors/destination-snowflake/src/test-integration/kotlin/io/airbyte/integrations/destination/snowflake/operation/SnowflakeStagingClientIntegrationTest.kt (hunk 2 above)
Comment (edgao) [reply airbyte-integrations/connectors/destination-snowflake/src/test-integration/kotlin/io/airbyte/integrations/destination/snowflake/operation/SnowflakeStagingClientIntegrationTest.kt:42]: something like that :shrug: not sure about the exact syntax, presumably there's some way to specify "please load into _exactly these columns_ in the target table"

Comment (edgao) [reply airbyte-integrations/connectors/destination-snowflake/src/test-integration/kotlin/io/airbyte/integrations/destination/snowflake/operation/SnowflakeStagingClientIntegrationTest.kt:42]: which I guess is actually slightly different from what you have (i.e. "read these columns out of the file" vs "write to these columns in the table")

File: airbyte-integrations/connectors/destination-snowflake/src/main/kotlin/io/airbyte/integrations/destination/snowflake/operation/SnowflakeStagingClient.kt (hunk 1 above)
Comment (gisripa) [reply airbyte-integrations/connectors/destination-snowflake/src/main/kotlin/io/airbyte/integrations/destination/snowflake/operation/SnowflakeStagingClient.kt:115]: @talnidam surely can use `minOf(..,99)` however there is no added benefit in the way we upload 1 file only. We use a 200M file and only 1 file per PUT and snowflake chunks it if > 64M which I'm guessing won't be more than 3 or 4 chunks.  The advantage of this shows up if we do PUT with a `directory/*` with many files in the directory. Also the default is 4 in snowflake settings if not provided. so just used that as max.

**Pull request merged**